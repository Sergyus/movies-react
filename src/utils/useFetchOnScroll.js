import { useLayoutEffect } from 'react';
import getScrollDownPercentage from './scroll';

export const handleScroll = (
  isLoading, hasMorePages, filter, currentPage,
  getLatestMovies, getTopRatedMovies, getUpcomingMovies,
  getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery, window,
) => {
  if (!isLoading) {
    const { scrollHeight, clientHeight } = window.document.documentElement;
    const scrollPos = window.pageYOffset;
    const percentageScrolled = getScrollDownPercentage(
      { scrollHeight, clientHeight, scrollPos },
    );
    if (percentageScrolled > 0.9 && hasMorePages) {
      switch (filter) {
        case 'Trending':
          getLatestMovies(currentPage + 1);
          break;
        case 'Top Rated':
          getTopRatedMovies(currentPage + 1);
          break;
        case 'Coming Soon':
          getUpcomingMovies(currentPage + 1);
          break;
        case 'Genres':
          getMoviesByGenre(currentPage + 1, currentGenre);
          break;
        case 'Search':
          getSearchMovies(currentPage + 1, currentSearchQuery);
          break;
        default: break;
      }
    }
  }
};
export const useFetchOnScroll = (...args) => {
  useLayoutEffect(() => {
    const callback = () => handleScroll(...args, window);
    document.addEventListener('scroll', callback);
    return () => {
      document.removeEventListener('scroll', callback);
    };
  }, [...args]);
};
