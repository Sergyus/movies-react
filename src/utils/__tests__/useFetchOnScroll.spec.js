import { JSDOM } from 'jsdom';
import React from 'react';
import { act, create } from 'react-test-renderer';
import { handleScroll, useFetchOnScroll } from '../useFetchOnScroll';

describe('Fetch on scroll work correctly', () => {
  const currentPage = 1;
  const currentGenre = 12;
  const currentSearchQuery = '123';
  const window = {
    document: {
      documentElement: {
        scrollHeight: 1600,
        clientHeight: 1200,
      },
    },
    pageYOffset: 500,
  };
  const getLatestMovies = jest.fn();
  const getTopRatedMovies = jest.fn();
  const getUpcomingMovies = jest.fn();
  const getMoviesByGenre = jest.fn();
  const getSearchMovies = jest.fn();
  it('Fetch on scroll if not load and filter trending', () => {
    const isLoading = false;
    const hasMorePages = true;
    const filter = 'Trending';
    handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery, window,
    );
    expect(getLatestMovies).toHaveBeenCalled();
  });
  it('Fetch on scroll if not load and filter top rated', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Top Rated';
    handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    );
    expect(getTopRatedMovies).toHaveBeenCalled();
  });
  it('Fetch on scroll if not load and filter coming soon', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Coming Soon';
    handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    );
    expect(getUpcomingMovies).toHaveBeenCalled();
  });
  it('Fetch on scroll if not load and filter genres', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Genres';
    handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    );
    expect(getMoviesByGenre).toHaveBeenCalled();
  });
  it('Fetch on scroll if not load and filter search', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Search';
    handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    );
    expect(getSearchMovies).toHaveBeenCalled();
  });
  it('Fetch on scroll if not load without filter', () => {
    const hasMorePages = true;
    const isLoading = false;
    const filter = 'Hello';
    const mockFn = jest.fn(() => handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    ));
    act(() => {
      mockFn();
    });
    expect(mockFn).toHaveBeenCalled();
  });
  it('Fetch on scroll if loading = true', () => {
    const hasMorePages = true;
    const isLoading = true;
    const filter = 'Hello';
    const mockFn = jest.fn(() => handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    ));
    act(() => {
      mockFn();
    });
    expect(mockFn).toHaveBeenCalled();
  });
  it('Fetch on scroll if hasn\'t more pages', () => {
    const hasMorePages = false;
    const isLoading = false;
    const filter = 'Hello';
    const mockFn = jest.fn(() => handleScroll(
      isLoading, hasMorePages, filter, currentPage,
      getLatestMovies, getTopRatedMovies, getUpcomingMovies,
      getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      window,
    ));
    act(() => {
      mockFn();
    });
    expect(mockFn).toHaveBeenCalled();
  });
  it('Callback call correctly', () => {
    const dom = new JSDOM();
    global.document = dom.window.document;
    global.window = dom.window;
    const hasMorePages = false;
    const isLoading = false;
    const filter = 'Hello';
    const Component = () => {
      useFetchOnScroll(
        isLoading, hasMorePages, filter, currentPage,
        getLatestMovies, getTopRatedMovies, getUpcomingMovies,
        getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
      );
      return <h1>Hello</h1>;
    };
    let tree;
    act(() => {
      tree = create(<Component />);
    });
    const event = new Event('scroll', {});
    act(() => {
      document.dispatchEvent(event);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
