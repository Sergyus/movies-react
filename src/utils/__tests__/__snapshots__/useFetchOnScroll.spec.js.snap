// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch on scroll work correctly Callback call correctly 1`] = `
<h1>
  Hello
</h1>
`;
